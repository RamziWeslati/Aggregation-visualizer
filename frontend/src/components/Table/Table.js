import React from "react";
import PropTypes from 'prop-types';

import {
  Table,
  TableBody,
  TableCell,
  TableContainer,
  TableHead,
  TableRow,
} from "@material-ui/core";

import { StyledTableCell } from "../../styles/Table.theme";

const SimpleTable = ({ data, header }) => {
  /**Presentational component that renders data in a row column fashion 
   * 
   * @props
   * - data: array of object
   *      data to display, each object is mapped to a row
   * - header [optional]: array of string
   *      header to display as column names. defaulted to keys of a data object
   */

  // column names definition
  const columns = header ? header : Object.keys(data[0]);

  return (
    <TableContainer style={{maxHeight: '300px'}}>
      <Table stickyHeader>
        <TableHead >
          <TableRow data-testid="header">
            {columns.map((column, idx) => (
              <StyledTableCell key={`header-cell-${idx}`}>{column}</StyledTableCell>
            ))}
          </TableRow>
        </TableHead>
        <TableBody>
          {data.flatMap((rowData, rIdx) => (
            <TableRow data-testid="row" key={`table-row-${rIdx}`}>
              {Object.values(rowData).map((rowValue, cIdx) => (
                <TableCell key={`table-row-${cIdx}-${rIdx}`}>{rowValue}</TableCell>
              ))}
            </TableRow>
          ))}
        </TableBody>
      </Table>
    </TableContainer>
  );
};


SimpleTable.propTypes = {
  header: PropTypes.array, // optional, defaulted in component using props
  
  // custom prop checker to require non empty arrays of data
  data: (props, propName, componentName) => {
    const errorBaseString = `Invalid prop ${propName} supplied to ${componentName}, `
    const dataProp = props[propName]

    if (dataProp === undefined)
      return new Error(
        errorBaseString + 'data prop is required.'
      );
    if (!Array.isArray(dataProp))
      return new Error(
        errorBaseString + 'data prop must be an array.'
      );
    if (dataProp.length == 0)
      return new Error(
        errorBaseString + 'data prop must not be empty.'
    );
  },
};

export default SimpleTable;
