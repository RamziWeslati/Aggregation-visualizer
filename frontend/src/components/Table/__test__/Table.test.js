import React from 'react';
import ReactDom from 'react-dom';

import { render } from '@testing-library/react';
import * as renderer from 'react-test-renderer';

import Table from './../Table';

const row1 = {'c11': 'v11', 'c12': 'v12', 'c13': 'v13'}
const row2 = {'c21': 'v21', 'c22': 'v22', 'c23': 'v23'}

const RequiredProps = {
    data: [row1, row2]
}

const optionalProps = {
    headers: ['Column Name 1', 'Column Name 2', 'Column Name 3']
}

const props = {...optionalProps, ...RequiredProps}

// cleanup is automated

// uts
it('renders without crashing, only required props', () => {
    const div = document.createElement('div')
    ReactDom.render(<Table {...RequiredProps} />, div)
})

it('renders without crashing, full props', () => {
    const div = document.createElement('div')
    ReactDom.render(<Table {...props} />, div)
})

it(`renders ${props.data.length} rows`, () => {
    const tree = render(<Table {...props} />)
    expect(tree.getAllByTestId('row')).toHaveLength(props.data.length)
})

it(`renders ${props.data.length + 1} row role`, () => { // header and data rows
    const tree = render(<Table {...props} />)
    expect(tree.getAllByRole('row')).toHaveLength(props.data.length + 1)  // + 1 for header row
})

it(`renders a header when header is passed`, () => {
    const tree = render(<Table {...props} />)
    expect(tree.getByTestId('header')).not.toBeNull()
})

it(`renders an infered header when header is not passed`, () => {
    const tree = render(<Table {...RequiredProps} />)
    expect(tree.getByTestId('header')).not.toBeNull()
})


// snapshots
it('matches infered header snapshot', () => {
    const tree = renderer.create(<Table {...RequiredProps} />).toJSON();
    expect(tree).toMatchSnapshot();
})

it('matches custom header snapshot', () => {
    const tree = renderer.create(<Table {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
})
