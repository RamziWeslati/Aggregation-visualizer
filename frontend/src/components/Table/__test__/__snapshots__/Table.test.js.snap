// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches custom header snapshot 1`] = `
<div
  className="MuiTableContainer-root"
  style={
    Object {
      "maxHeight": "300px",
    }
  }
>
  <table
    className="MuiTable-root MuiTable-stickyHeader"
    role={null}
  >
    <thead
      className="MuiTableHead-root"
      role={null}
    >
      <tr
        className="MuiTableRow-root MuiTableRow-head"
        data-testid="header"
        role={null}
      >
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head WithStyles(ForwardRef(TableCell))-head-1 MuiTableCell-stickyHeader"
          scope="col"
        >
          c11
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head WithStyles(ForwardRef(TableCell))-head-1 MuiTableCell-stickyHeader"
          scope="col"
        >
          c12
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head WithStyles(ForwardRef(TableCell))-head-1 MuiTableCell-stickyHeader"
          scope="col"
        >
          c13
        </th>
      </tr>
    </thead>
    <tbody
      className="MuiTableBody-root"
      role={null}
    >
      <tr
        className="MuiTableRow-root"
        data-testid="row"
        role={null}
      >
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        >
          v11
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        >
          v12
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        >
          v13
        </td>
      </tr>
      <tr
        className="MuiTableRow-root"
        data-testid="row"
        role={null}
      >
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        >
          v21
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        >
          v22
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        >
          v23
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`matches infered header snapshot 1`] = `
<div
  className="MuiTableContainer-root"
  style={
    Object {
      "maxHeight": "300px",
    }
  }
>
  <table
    className="MuiTable-root MuiTable-stickyHeader"
    role={null}
  >
    <thead
      className="MuiTableHead-root"
      role={null}
    >
      <tr
        className="MuiTableRow-root MuiTableRow-head"
        data-testid="header"
        role={null}
      >
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head WithStyles(ForwardRef(TableCell))-head-1 MuiTableCell-stickyHeader"
          scope="col"
        >
          c11
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head WithStyles(ForwardRef(TableCell))-head-1 MuiTableCell-stickyHeader"
          scope="col"
        >
          c12
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head WithStyles(ForwardRef(TableCell))-head-1 MuiTableCell-stickyHeader"
          scope="col"
        >
          c13
        </th>
      </tr>
    </thead>
    <tbody
      className="MuiTableBody-root"
      role={null}
    >
      <tr
        className="MuiTableRow-root"
        data-testid="row"
        role={null}
      >
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        >
          v11
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        >
          v12
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        >
          v13
        </td>
      </tr>
      <tr
        className="MuiTableRow-root"
        data-testid="row"
        role={null}
      >
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        >
          v21
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        >
          v22
        </td>
        <td
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-body"
        >
          v23
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
